gpbft/participant.go

```diff
@@ -66,63 +68,62 @@ func (p *Participant) Start() (err error) {
 }
 
 func (p *Participant) CurrentRound() uint64 {
-	if p.granite == nil {
+	if p.gpbft == nil {
 		return 0
 	}
-	return p.granite.round
+	return p.gpbft.round
 }
 
-// Validates a message received from another participant, if possible.
-// An invalid message can never become valid, so may be dropped.
-// A message can only be validated if it is for the currently-executing protocol instance.
-// Returns whether the message could be validated, and an error if it was invalid.
-func (p *Participant) ValidateMessage(msg *GMessage) (checked bool, err error) {
+// Receives a Granite message from some other participant.
+//   - A message for a previous instance is dropped.
+//   - A message for the current instance is validated and delivered immediately.
+//   - A message for a future instance for which a committee is available is validated and
+//     locally queued, to be delivered when that instance begins.
+//   - A message for a future instance for which a committee is not available returns an error.
+//   - An invalid message returns an error.
+//
+// Returns whether the message was accepted for the current instance or a near future instance,
+// and/or an error if it was invalid or could not be processed.
+func (p *Participant) ReceiveMessage(msg *GMessage) (accepted bool, err error) {
```

Comment:
Please give feedback on the return values here. Errors could come either from message validation-related things, which can pretty safely be ignored, or from other more serious sources. The boolean return is relevant whether or not error is nil. Roughly, if false then errors are inconsequential, and if true then errors are serious. But this breaks the usual pattern of ignoring return value with non-nil errors.

One alternative would be to return `(validation error, err error)` two possibly-nil errors, a traditional one and one indicating validation results.

Reply:
Another option would be to have a sentinel error value which gets wrapped with a cause.